### ISSUE TITLE
Make the selection tables context sensitive

### ISSUE THREAD

**Issue by aksajhfduwafe** [aksajhfduwafe/bugzilla-to-github-issues-test#1698]
| Field | Value |
|---|---|
| **Reporter** | @LennartH |
| **Assignee** | @LennartH |
| **Product** | Sailing Race Analytics |
| **Component** | DataMining |
| **Version** | unspecified |
| **Hardware** | PC |
| **OS** | Windows |
| **Severity** | normal |
| **Priority** | P2 |
| **Created** | 2014-01-16T07:23:29Z |
| **Original Bug** | [Bug 1698](https://bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=1698) |

---

The available values of the selection tables should depend on the selection.
For example: If you filter for a boat class/race only competitors should be available, who sail with the boat class/sailed in the race.

**Comment by aksajhfduwafe** [comment]
**@LennartH** commented:

Implemented a simple context sensitivity in the way, that new queries for the dimension values use the current filter selection. This was done with commit c9c50faca4fc2da3765acae08704d3456a9299fa on branch dataMining.

I noted a problem with filtered dimension queries. For example is it possible to retrieve data with no usable statistic values, that will be included in the dimension values. This looks like, the filtration isn't working. A concrete example can be seen in the data mining of the single race 505 race 1 of the Kieler Woche 2013:
- Calculate the Distance Traveled in m based on HasTrackedLegOfCompetitorContext, grouped by Leg Number and Competitor Team Name
- You can see that the competitor Thure Gnadeberg DEN 9062 has no data in the legs 2, 3, 4 and 5
- Select the Leg Number 2 on the level Leg, to filter the data
- Display the Competitor Boat Sail Number and/or the Competitor Team Name on the level Leg of competitor
- You can find Thure Gnadeberg and/or DEN 9062 in the displayed data
This is because the competitor is listed in Leg number 2, so the TrackedLegOfCompetitor for Leg 2 and the Competitor Thure Gnadeberg is retrieved, but it returns null for the statistic Distance Traveled in m. I think this problem is related to #1682.

The next step will be to implement, that new filter selections have an effect on already selected value filters. It could be hard to handle the possibilities of side effects. For example has a new filter selection a CHANGING effect of an already selected value filter, which will again have an effect on the already selected value filters.

**Comment by aksajhfduwafe** [comment]
**@LennartH** commented:

Done with commit 67175132eaada033ac4dc7c741c237bf76f2278e on branch dataMining.

This was achieved in multiple steps:
- At first I implemented an mechanism to reduce the displayed filter dimensions with commit 9c7eb03507ccd41192e036688ff04ca885d36933. This reduces the dimensions in a way, that dimensions of a higher retriever level aren't available again in lower levels. Dimensions are logically equal, if the display name, the parameter types (names) and the return type (name) are equal.
- This guarantees, that selection changes of level x can't effect the available data of tables of level y, where y < x.
- This allows to implement dynamic context sensitive selection tables in a simple way:
1. Start with the level, where the selection changed.
2. Update the tables and check, if the selection has been changed because of the update (previously selected items can be unavailable after the update and won't be selected anymore).
2.1 If the slection wasn't changed: Continue with the next table (or level)
2.2 Else: Start again with the current level.
3. Notify the listeners, that the selection has been changed.
